packages/odyssey-react-mui/src/Autocomplete.tsx

```diff
@@ -371,60 +396,92 @@ const Autocomplete = <
-    const overscanRowCount = 8;
+    const getListBoxHeight = useCallback(() => {
+      // 8px of padding top/bottom applied by MUI
+      const LISTBOX_PADDING = 8;
+
+      if (itemData.length > OVERSCAN_ROW_COUNT) {
+        // has a max-height of 40vh set in CSS. This is only set because height needs to be a number
+        return 1000;
+      } else {
+        const itemsHeightCalculated = itemData
+          .map((_, index) => sizeMap.current[index] || 0)
+          .reduce((a, b) => a + b, 0);
+        return LISTBOX_PADDING * 2 + itemsHeightCalculated;
+      }
+    }, [itemData, sizeMap]);
 
-    const itemSize = useCallback(() => optionHeight, []);
+    useEffect(() => {
+      if (sizeMap.current.length && itemData.length) {
+        setListHeight(getListBoxHeight());
+      }
+    }, [getListBoxHeight, itemData, sizeMap]);
+
+    // The number of items (rows or columns) to render outside of the visible area for performance and scrolling reasons
+    const OVERSCAN_ROW_COUNT = 8;
 
     const gridRef = useResetCache(itemData.length);
 
-    const renderWindow = useCallback(
-      ({ height, width }: AutoSizerSize) => (
-        <VariableSizeList
-          innerElementType="ul"
-          itemData={itemData}
-          itemCount={itemData.length}
-          // eslint-disable-next-line @typescript-eslint/no-unused-vars
-          itemSize={itemSize}
-          height={height}
-          width={width}
-          ref={gridRef}
-          outerElementType={OuterListboxElementType}
-          overscanCount={overscanRowCount}
-        >
-          {renderVirtualizedRow}
-        </VariableSizeList>
-      ),
-      [itemData, gridRef, itemSize],
+    const setItemSize = useCallback<SetItemSize>(
+      (size) => {
+        gridRef?.current?.resetAfterIndex(0, true);
+        sizeMap.current = sizeMap.current.concat(size);
+      },
+      [gridRef, sizeMap],
+    );
+    const getItemSize = useCallback(
+      // using 45px as a sane default here to avoid a lot of content shift on repaint
+      (index: number) => sizeMap.current[index] || 45,
+      [sizeMap],
     );
 
     return (
       <ListboxContainer ref={ref}>
         <OuterListboxContext.Provider value={other}>
-          <AutoSizer>{renderWindow}</AutoSizer>
+          <VariableSizeList
+            innerElementType="ul"
+            itemData={itemData}
+            itemCount={itemData.length}
+            itemSize={getItemSize}
+            height={listHeight}
+            width="100%"
+            ref={gridRef}
+            outerElementType={OuterListboxElementType}
+            overscanCount={OVERSCAN_ROW_COUNT}
+          >
+            {({ data, index, style }) => (
+              <Row
+                data={data}
+                index={index}
+                style={style}
+                setItemSize={setItemSize}
+              />
+            )}
+          </VariableSizeList>
```

Comment:
Don't we need a `key` prop on these rows?

Preferably one where it's `key={data.value}`, so it improves performance.

Reply:
@KevinGhadyani-Okta The underlying `li` component from MUI already has a key, I pull it out and apply it [here](https://github.com/okta/odyssey/pull/2321/files/bb8387c4169503e865b98a84507f84ac67433330#diff-3acb2db7ecf79128f9dbd0ff306e83ddef79cefce67a12a3fa58f7fa2d4dff75R385)

Reply:
That won't work. The `key` is something relevant to the scope of a loop. It doesn't matter if it's one level deep in a sub-component.

It should be defined at this level:
Source: https://legacy.reactjs.org/docs/lists-and-keys.html#extracting-components-with-keys
![image](https://github.com/user-attachments/assets/9cdc6c1c-1ea0-4aa1-86d6-378cde197640)